core/src/main/java/io/questdb/cairo/CairoMetadata.java

```diff
@@ -0,0 +1,613 @@
+                if (cachedTable == null) {
+                    localCache.put(latestTable.getTableName(), latestTable);
+                } else if (cachedTable.getMetadataVersion() < latestTable.getMetadataVersion()) {
+                    localCache.put(cachedTable.getTableName(), latestTable);
+                } else if (cachedTable.getMetadataVersion() > latestTable.getMetadataVersion()) {
+                    throw new RuntimeException("disordered metadata versions");
+                } else {
+                    assert cachedTable.getMetadataVersion() == latestTable.getMetadataVersion();
+                    // otherwise its up to date, so we loop
+                }
+            }
+
+            iterator = localCache.entrySet().iterator();
+            while (iterator.hasNext()) {
+                CairoTable cachedTable = iterator.next().getValue();
+                CairoTable latestTable = tables.get(cachedTable.getTableName());
+                if (latestTable == null) {
+                    // if its not in the main cache, removed it from local cache
+                    iterator.remove();
+                } else if (cachedTable.getMetadataVersion() < latestTable.getMetadataVersion()) {
+                    localCache.put(cachedTable.getTableName(), latestTable);
+                } else if (cachedTable.getMetadataVersion() > latestTable.getMetadataVersion()) {
+                    throw new RuntimeException("disordered metadata versions");
+                } else {
+                    assert cachedTable.getMetadataVersion() == latestTable.getMetadataVersion();
+                    // otherwise its up to date, so we loop
+                }
+            }
+
+            return version;
+        }
+
+        /**
+         * For debug printing the metadata object, doesn't require manual closing.
+         */
+        @TestOnly
+        public String toString0() {
+            StringSink sink = tlSink.get();
+            sink.put("CairoMetadata [");
+            sink.put("tableCount=").put(tables.size()).put(']');
+            sink.put('\n');
+
+            for (CairoTable table : tables.values()) {
+                sink.put('\t');
+                table.toSink(sink);
+                sink.put('\n');
+            }
+            close();
+            String s = sink.toString();
+            sink.clear();
+            tlSink.remove();
+            return s;
+        }
+    }
+
+    private class CairoMetadataWriter extends CairoMetadataReader implements Closeable, CairoMetadataRW, AutoCloseable {
+
+        public void clear() {
+            tables.clear();
+        }
+
+        @Override
+        public void close() {
+            lock.writeLock().unlock();
+        }
+
+        public void dropTable(@NotNull CharSequence tableName) {
+            tables.remove(tableName);
+            LOG.info().$("dropped metadata [table=").$(tableName).I$();
+        }
+
+        public void dropTable(@NotNull TableToken tableToken) {
```

Comment:
nit: override methods should have @Override annotation - to be consistent with the rest of the codebase